### PR TITLE
UCP/AM/API: Define data fetch request and reply

### PR DESCRIPTION
**Pull request by brminich**
## What
Extend UCP AM API to allow requesting data from the target.

## Why ?
Mainly performance reasons.
We have IODEMO example demonstrating UCX usage in various IO frameworks. This API would help to implement READ requests

## How ?
- Add a specific flag which would tell `ucp_am_send_nbx` to request data from the target and store it in its buffer (provided in ucp_request_param_t.reply_buffer).
- Add new `ucp_am_send_reply_nbx`, which would accept `data_desc` (like `ucp_am_recv_nbx`) and send data back to the requestor.

### CODE REVIEW

**Comment by brminich** [comment]
UCP AM fetch

**Comment by brminich** [comment]
@yosefe, @shamisp, can you please take a look?

**Comment by shamisp** [comment]
@brminich In "theory" we should be able to send reply using existing APIs (put or send on a dedicated tag ?).

**Comment by shamisp** [comment]
Seems like in your example you don't invoke the reply function from the callback ? What happen with request generated by the reply function inside of the callback ?

**Comment by brminich** [comment]
> Seems like in your example you don't invoke the reply function from the callback ? What happen with request generated by the reply function inside of the callback ?

In the example reply function is called from the callback `am_recv_cb`. The reply function returns `ucs_status_ptr_t`, like other sending functions. I'll update the picture

**Comment by brminich** [comment]
> i



> @brminich In "theory" we should be able to send reply using existing APIs (put or send on a dedicated tag ?).

Yes, but handling different corner cases (data types, memory types, etc) would require quite many code to be written by the user. With this API handling of these cases will be done by UCP

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -488,7 +488,7 @@ enum ucp_send_am_flags {
                                               for AM sends. */
     UCP_AM_SEND_RNDV       = UCS_BIT(2), /**< Force UCP to use only rendezvous protocol
                                               for AM sends. */
-    UCP_AM_SEND_FETCH_DATA = UCS_BIT(3)  /**< Tells @ref ucp_am_send_nbx to send
+    UCP_AM_SEND_GET_REPLY  = UCS_BIT(3)  /**< Tells @ref ucp_am_send_nbx to send
```
Comment (yosefe) [inline src/ucp/api/ucp.h:491]: write which fields in request_param have to be specified (reply buffer, reply count)

**Review by yosefe — APPROVED** [review]

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (brminich) [reply src/ucp/api/ucp.h:491]: fixed

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -490,9 +490,9 @@ enum ucp_send_am_flags {
                                               for AM sends. */
     UCP_AM_SEND_GET_REPLY  = UCS_BIT(3)  /**< Tells @ref ucp_am_send_nbx to send
                                               request for data. In this case reply
-                                              buffer must be specified in
-                                              @ref ucp_request_param_t passed to
-                                              @ref ucp_am_send_nbx. Also,
+                                              buffer and reply count must be
+                                              specified in @ref ucp_request_param_t
+                                              passed to @ref ucp_am_send_nbx. Also,
```
Comment (yosefe) [inline src/ucp/api/ucp.h:495]: pls add ref to the exact fields

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -2863,8 +2861,8 @@ ucs_status_ptr_t ucp_am_send_nbx(ucp_ep_h ep, unsigned id,
  *                                the application is responsible for releasing
  *                                the handle using @ref ucp_request_free routine.
  */
-ucs_status_ptr_t ucp_am_send_reply_nbx(ucp_ep_h ep, void *data_desc,
-                                       const void *buffer, size_t count,
+ucs_status_ptr_t ucp_am_send_reply_nbx(void *data_desc, const void *buffer,
```
Comment (yosefe) [inline src/ucp/api/ucp.h:2864]: maybe pass worker?

**Pull request closed without merging**